### DISCUSSION TITLE
Key rotations when using TCE

### DISCUSSION THREAD

**Discussion by vidbina** [supabase/supabase#16379] [Questions]
Hi y'all, just looking for a friend 😅 

# Background

Am currently using Supabase for a build and was reading into TCE to handle secrets management (like API keys and data source secrets) while enjoying row-level encryption. While thinking through the design, I couldn't quite figure out how key rotations are managed in Supabase when using TCE.

## Resources consulted:
- https://supabase.com/blog/transparent-column-encryption-with-postgres#one-key-id-per-row-with-nonce-support
- https://supabase.com/blog/supabase-vault
- https://github.com/michelp/pgsodium

# Question

If we a) update a pgsodium key or b) change the `key_id` field for a row that uses a key (TCE),
1. will affected cipher texts be recomputed (i.e.: decrypted against old key and encrypted against new key), 
2. is there a specific `SECURITY LABEL` statement that needs to be executed or 
3. **do we need to implement key rotation ourselves**?

# Working understanding

1. No, affected cipher texts are not recomputed by pgsodium as hinted by this passage from https://github.com/michelp/pgsodium#server-key-derivation
   > Key rotation is up to you, whatever strategy you want to go from one key to the next. A simple strategy is incrementing the key id and re-encrypting from N to N+1. Newer keys will have increasing ids, you can always tell the order in which keys are superceded.
2. Not AFAIK
3. Yes to DIY, based on the quote listed in point 1

So AFAIK, key rotation (at this level) is not a thing and end-users will have to manage this entirely themselves.

## Follow-up question

In case my understanding is correct, does Supabase have some best-practices in mind for users to consider in implementing key rotation?

# TL;DR: What I tried to verify that key rotation isn't a thing (in case you'd like to try for yourself)

Here is an experiment, I ran to test things (replace magic strings for your own values, I'm using these in a local Supabase setup just for testing so forgive my laziness in not redacting these out):

1.  Stub table
    
    ```sql
    CREATE TABLE "public"."secrets_sandbox" (
    "id" uuid NOT NULL PRIMARY KEY DEFAULT gen_random_uuid(),
    "key_id" uuid NOT NULL,
    "nonce" bytea DEFAULT pgsodium.crypto_aead_det_noncegen(),
    "secret" text
    );
    ```

5.  Insert dummy record using a key previously created using the `pgsodium.create_key()` function
    
    ```sql
    INSERT INTO "public"."secrets_sandbox" (key_id, secret)
    VALUES ('8342c975-afe7-4be8-b750-2a0008f2b070', 'Hello World!');
    ```

6.  Update entry just to confirm that updates happen in plain text
    
    ```sql
    UPDATE "secrets_sandbox"
    SET secret = 'Something else'
    WHERE id = '57618333-1f95-41c1-adb6-e91f80f10bba'
    RETURNING secret, id AS updated_secret;
    ```

7.  Set security label to observe which changes happen on table
    
    ```sql
    SECURITY LABEL FOR pgsodium
    ON COLUMN secrets_sandbox.secret
    IS 'ENCRYPT WITH KEY COLUMN key_id NONCE nonce';
    ```

8.  Observe that the table row did not change

9.  Run another update in order to trigger encryption and **observe that the secret field is no longer in plaintext**
    
    ```sql
    UPDATE "secrets_sandbox"
    SET secret = 'This should be hidden'
    WHERE id = '57618333-1f95-41c1-adb6-e91f80f10bba'
    RETURNING secret, id AS updated_secret;
    ```

10. Read the decrypted text from the `decrypted_secrets_sandbox` view:
    
    ```sql
    SELECT decrypted_secret, secret FROM decrypted_secrets_sandbox
    ```

11. Create another key and update the record again to observe if the secret will be reencrypted
    
    ```sql
    UPDATE "secrets_sandbox"
    SET key_id = (SELECT id FROM pgsodium.create_key())
    WHERE id = '57618333-1f95-41c1-adb6-e91f80f10bba'
    RETURNING secret, id AS updated_secret;
    ```

12. Attempt to decrypt the secret again
    
    ```sql
    SELECT decrypted_secret, secret FROM decrypted_secrets_sandbox
    ```
    
    and observe the error.
    
    ```sql
    pgsodium_crypto_aead_det_decrypt_by_id: invalid ciphertext
    ```
    
    Conclusion: key changes do not trigger reencryption. Key rotation needs a different configuration or needs to be implemented by the developer!

    Update: When using the `ASSOCIATED` clause  to define the security label, the same holds true. An update of the associated fields does not automatically recompute the ciphertext which is why changes to the associated fields will break the encryption (until the associated fields are restored to the value they had when the original ciphertext was computed).

13. Wrap up the experiment by removing the test table
    
    ```sql
    DROP TABLE "public"."secrets_sandbox" CASCADE;
    ```

**Comment by vidbina** [comment]
Just responding to an ongoing Discord convo https://discordapp.com/channels/839993398554656828/1138834134056308908, tried the following steps.

Try them step-by-step and 
1. update the key id for the output you get in step 1 (you only need it in step 2 when we prep the table).
2. replace `SOME_USER_EMAIL` for a valid e-mail address that you're trying to impersonate during debug 

```sql
-- 1
select * from pgsodium.create_key()
-- RESULT: d3d36a8f-c7f0-4bac-ad22-fd61076143c5

-- 2
CREATE TABLE credit_cards (
	id bigserial primary key,
	credit_card_number text
);

SECURITY LABEL FOR pgsodium
	ON COLUMN credit_cards.credit_card_number
	IS 'ENCRYPT WITH KEY ID d3d36a8f-c7f0-4bac-ad22-fd61076143c5';

-- 3
insert into credit_cards
  (credit_card_number)
values
  ('1234-5678-8765-4321');
-- RESULT: id 1

-- 4 Stub helper functions for impersonation, login as vid@fortoo and view role
grant anon, authenticated to postgres;

create or replace procedure auth.login_as_user (user_email text)
    language plpgsql
    as $$
declare
    auth_user auth.users;
begin
    select
        * into auth_user
    from
        auth.users
    where
        email = user_email;
    execute format('set request.jwt.claim.sub=%L', (auth_user).id::text);
    execute format('set request.jwt.claim.role=%I', (auth_user).role);
    execute format('set request.jwt.claim.email=%L', (auth_user).email);
    execute format('set request.jwt.claims=%L', json_strip_nulls(json_build_object('app_metadata', (auth_user).raw_app_meta_data))::text);

    raise notice '%', format( 'set role %I; -- logging in as %L (%L)', (auth_user).role, (auth_user).id, (auth_user).email);
    execute format('set role %I', (auth_user).role);
end;
$$;

create or replace procedure auth.login_as_anon ()
    language plpgsql
    as $$
begin
    set request.jwt.claim.sub='';
    set request.jwt.claim.role='';
    set request.jwt.claim.email='';
    set request.jwt.claims='';
    set role anon;
end;
$$;

create or replace procedure auth.logout ()
    language plpgsql
    as $$
begin
    set request.jwt.claim.sub='';
    set request.jwt.claim.role='';
    set request.jwt.claim.email='';
    set request.jwt.claims='';
    set role postgres;
end;
$$;

call auth.login_as_user(SOME_USER_EMAIL);
SELECT current_role;
-- RESULT: authenticated

-- 5 Change to authenticated user and insert new record
call auth.login_as_user(SOME_USER_EMAIL);
insert into credit_cards
  (credit_card_number)
values
  ('1234-5678-8765-4321');
-- RESULT: Failed to run sql query: permission denied for function crypto_aead_det_encrypt

-- 6 Retry with grant preceding
GRANT pgsodium_keyiduser TO authenticated;
call auth.login_as_user(SOME_USER_EMAIL);
insert into credit_cards
  (credit_card_number)
values
  ('1234-5678-8765-4321');
-- RESULT: permission denied for function crypto_aead_det_encrypt
```

**Reply by vidbina** [reply]
@dijonmusters if you have some headspace, roasting the errors in my thinking here would be appreciated. 🙏🏿  My current understanding is that a) key rotation is a DIY concern and b) the Supabase `service_role` (post PG `postgres`) roles must be assumed in order to call pgsodium funcs.
